File: app/src/main/java/com/nfcalarmclock/NacSnackbar.java
package com.nfcalarmclock;

import android.content.Context;
import android.view.View;
import java.util.LinkedList;
import java.util.Queue;
import com.google.android.material.snackbar.BaseTransientBottomBar;
import com.google.android.material.snackbar.Snackbar;

/**
 */
public class NacSnackbar
	implements View.OnClickListener
{

	/**
	 * Snackbar data holder.
	 */
	private static class SnackHolder
	{

		/**
		 * Message.
		 */
		private final String message;

		/**
		 * Action button text.
		 */
		private final String action;

		/**
		 * Click listener.
		 */
		private final View.OnClickListener onClickListener;

		/**
		 */
		public SnackHolder(String message, String action,
			View.OnClickListener onClickListener)
		{
			this.message = message;
			this.action = action;
			this.onClickListener = onClickListener;
		}

	}

	/**
	 * Response to snackbar onDismissed and onShown callbacks.
	 */
	private class NacSnackbarCallback
		extends BaseTransientBottomBar.BaseCallback<Snackbar>
	{

		@Override
		public void onDismissed(Snackbar transientBottomBar, int event)
		{
			Queue<SnackHolder> queue = getQueue();
			mSnackbar = null;

			if (queue.size() > 0)
			{
				SnackHolder holder = queue.remove();
				String message = holder.message;
				String action = holder.action;
				View.OnClickListener onClickListener = holder.onClickListener;

				show(message, action, onClickListener, false);
			}
		}

		@Override
		public void onShown(Snackbar transientBottomBar)
		{
		}

	}

	/**
	 * Snackbar.
	 */
	private Snackbar mSnackbar;

	/**
	 * Root view.
	 */
	private final View mRoot;

	/**
	 * Message queue.
	 */
	private final Queue<SnackHolder> mQueue;

	/**
	 * Check if the snackbar can be dismissed early.
	 */
	private boolean mCanDismiss;

	/**
	 */
	public NacSnackbar(View root)
	{
		this.mSnackbar = null;
		this.mRoot = root;
		this.mQueue = new LinkedList<>();
		this.mCanDismiss = false;
	}

	/**
	 * Check if the snackbar can be dismissed early.
	 */
	public boolean canDismiss()
	{
		return this.mCanDismiss;
	}

	/**
	 * Dismiss the snackbar.
	 */
	public void dismiss()
	{
		Snackbar snackbar = this.getSnackbar();

		if (snackbar != null)
		{
			snackbar.dismiss();
		}
	}

	/**
	 * @return The context.
	 */
	public Context getContext()
	{
		View root = this.getRoot();
		return (root != null) ? root.getContext() : null;
	}

	/**
	 * @return The root view.
	 */
	public View getRoot()
	{
		return this.mRoot;
	}

	/**
	 * @return The queue.
	 */
	private Queue<SnackHolder> getQueue()
	{
		return this.mQueue;
	}

	/**
	 * @return The snackbar.
	 */
	public Snackbar getSnackbar()
	{
		return this.mSnackbar;
	}

	/**
	 * @return True if the snackbar is shown, and False otherwise.
	 */
	public boolean isShown()
	{
		Snackbar snackbar = this.getSnackbar();
		return ((snackbar != null) && snackbar.isShown());
	}

	/**
	 * Default listener when the action button is clicked.
	 */
	@Override
	public void onClick(View view)
	{
	}

	/**
	 * Queue the snackbar data.
	 */
	private void queue(String message, String action,
		View.OnClickListener listener)
	{
		SnackHolder holder = new SnackHolder(message, action, listener);
		this.mQueue.add(holder);
	}

	/**
	 * Set the snackbar action listener.
	 */
	public void setAction(String action, View.OnClickListener listener)
	{
		if (action.isEmpty())
		{
			return;
		}

		if (listener == null)
		{
			listener = this;
		}

		Snackbar snackbar = this.getSnackbar();
		snackbar.setAction(action, listener);
	}

	/**
	 * Set the action text color to the theme color.
	 */
	public void setActionTextThemeColor()
	{
		Snackbar snackbar = this.getSnackbar();
		Context context = this.getContext();
		NacSharedPreferences shared = new NacSharedPreferences(context);
		int themeColor = shared.getThemeColor();

		snackbar.setActionTextColor(themeColor);
	}

	/**
	 * Set the snackbar.
	 */
	protected void setSnackbar(String message)
	{
		View root = this.getRoot();
		Snackbar snackbar = Snackbar.make(root, message, Snackbar.LENGTH_LONG);
		this.mSnackbar = snackbar;

		snackbar.addCallback(new NacSnackbarCallback());
	}

	/**
	 * Show the snackbar now.
	 */
	public void show()
	{
		this.getSnackbar().show();
	}

	/**
	 * @see #show(String, String, View.OnClickListener, boolean)
	 */
	@SuppressWarnings("unused")
	public void show(String message, String action,
		View.OnClickListener listener)
	{
		this.show(message, action, listener, false);
	}

	/**
	 * Setup the snackbar and show it.
	 *
	 * @see #show()
	 */
	public void show(String message, String action,
		View.OnClickListener listener, boolean dismiss)
	{
		this.mCanDismiss = dismiss;

		if (this.isShown())
		{
			this.dismiss();
			this.queue(message, action, listener);
			return;
		}

		this.setSnackbar(message);
		this.setAction(action, listener);
		this.setActionTextThemeColor();
		this.show();
	}

}

File: app/src/main/java/com/nfcalarmclock/view/snackbar/NacSnackbar.kt
package com.nfcalarmclock.view.snackbar

import android.content.Context
import android.view.View
import com.google.android.material.snackbar.BaseTransientBottomBar.BaseCallback
import com.google.android.material.snackbar.Snackbar
import com.nfcalarmclock.shared.NacSharedPreferences
import java.util.LinkedList
import java.util.Queue

/**
 * Snackbar handler to show, swipe away, etc.
 *
 * TODO: Can I get rid of this?
 */
class NacSnackbar(

	/**
	 * Root view.
	 */
	val root: View

) : View.OnClickListener
{

	/**
	 * Snackbar data holder.
	 */
	private class SnackHolder(

		/**
		 * Message.
		 */
		val message: String,

		/**
		 * Action button text.
		 */
		val action: String,

		/**
		 * Click listener.
		 */
		val onClickListener: View.OnClickListener?)

	/**
	 * Response to snackbar onDismissed and onShown callbacks.
	 */
	private inner class NacSnackbarCallback : BaseCallback<Snackbar?>()
	{

		/**
		 * Called when the snackbar is dismissed.
		 */
		override fun onDismissed(transientBottomBar: Snackbar?, event: Int)
		{
			val queue: Queue<SnackHolder> = queue
			snackbar = null

			// Check if there is a snackbar to show
			if (queue.size > 0)
			{
				val holder = queue.remove()
				val message = holder.message
				val action = holder.action
				val onClickListener = holder.onClickListener

				// Show the snackbar
				show(message, action, onClickListener, false)
			}
		}

		/**
		 * Called when the snackbar is shown.
		 */
		override fun onShown(transientBottomBar: Snackbar?)
		{
		}

	}

	/**
	 * Context.
	 */
	val context: Context
		get() = root.context

	/**
	 * Snackbar.
	 */
	var snackbar: Snackbar? = null
		private set

	/**
	 * Message queue.
	 */
	private val queue: Queue<SnackHolder> = LinkedList()

	/**
	 * Check if the snackbar can be dismissed early.
	 */
	var canDismiss = false

	/**
	 * Check if the snackbar is shown.
	 */
	val isShown: Boolean
		get() = snackbar?.isShown == true

	/**
	 * Dismiss the snackbar.
	 */
	fun dismiss()
	{
		snackbar?.dismiss()
	}

	/**
	 * Default listener when the action button is clicked.
	 */
	override fun onClick(view: View)
	{
	}

	/**
	 * Queue the snackbar data.
	 */
	private fun queue(message: String, action: String,
		listener: View.OnClickListener?)
	{
		val holder = SnackHolder(message, action, listener)

		queue.add(holder)
	}

	/**
	 * Set the snackbar action listener.
	 */
	fun setAction(action: String, listener: View.OnClickListener?)
	{
		if (action.isEmpty())
		{
			return
		}

		// Get the snackbar and listener
		val snackbar = snackbar
		val actionListener = listener ?: this

		// Set the action
		snackbar!!.setAction(action, actionListener)
	}

	/**
	 * Set the action text color to the theme color.
	 */
	fun setActionTextThemeColor()
	{
		val shared = NacSharedPreferences(context)
		val themeColor = shared.themeColor

		// Set the text color
		snackbar!!.setActionTextColor(themeColor)
	}

	/**
	 * Set the snackbar.
	 */
	protected fun setSnackbar(message: String?)
	{
		// Set the snackbar
		snackbar = Snackbar.make(root, message!!, Snackbar.LENGTH_LONG)

		// Add a callback
		snackbar!!.addCallback(NacSnackbarCallback())
	}

	/**
	 * Show the snackbar now.
	 */
	fun show()
	{
		snackbar!!.show()
	}

	/**
	 * @see .show
	 */
	@Suppress("unused")
	fun show(message: String, action: String,
		listener: View.OnClickListener?)
	{
		this.show(message, action, listener, false)
	}

	/**
	 * Setup the snackbar and show it.
	 *
	 * @see .show
	 */
	fun show(message: String, action: String,
		listener: View.OnClickListener?, dismiss: Boolean)
	{
		// Set whether the snackbar can be dismissed or not
		canDismiss = dismiss

		// Check if a snackbar is currently shown
		if (isShown)
		{
			// Dismiss the shown snackbar
			dismiss()

			// Queue the next one
			queue(message, action, listener)
			return
		}

		// Set the snackbar
		setSnackbar(message)

		// Setup the snackbar
		setAction(action, listener)
		setActionTextThemeColor()

		// Show the snackbar
		this.show()
	}

}